libbeat/esleg/eslegclient/connection_test.go

```diff
@@ -104,9 +104,35 @@ func TestHeaders(t *testing.T) {
 		require.NoError(t, err)
 
 		require.Equal(t, req.Header, http.Header(td.expected))
+
 	}
 }
 
+func TestUserAgentHeader(t *testing.T) {
+	conn, err := NewConnection(ConnectionSettings{
+		Beatname:         "testbeat",
+		UserAgentPostfix: "Agent",
+	})
+	require.NoError(t, err)
+
+	req, err := http.NewRequestWithContext(context.Background(), "GET", "http://localhost/some/path", nil)
+	require.NoError(t, err)
+
+	rawClient, ok := conn.HTTP.(*http.Client)
+	require.True(t, ok)
+
+	// don't want to check the error, since the URL doesn't exist, so we'll always return an error.
+	resp, _ := rawClient.Transport.RoundTrip(req)
+	// just to make linter happy, there's no body
+	if resp != nil {
+		resp.Body.Close()
+	}
+
+	// the `RoundTrip` will set the header in our request
+	require.Contains(t, req.Header.Get("User-Agent"), "Elastic-testbeat-Agent")
```

Comment:
While this tests the feature, it relies on the RoundTrip modifying the request.

I believe a more reliable way to test is to start an `httptest.Server`, make a request to it, and on it's handler check the User-Agent presented. This way we're testing the expected final behaviour and not just our implementation.

Reply:
hmm, yeah, should probably switch over to using https://github.com/elastic/mock-es

Reply:
The change did cover part of my comment/request of not relying on the modification of the request. However relying on `mock-es` and how it interacts with the metrics registry seems a bit over-complicated and risky in case `mock-es` changes it's implementation.

I had something simple in mind that relies only in the Go standard library:
```go
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		prefix := "Elastic-testbeat-Agent"
		if !strings.HasPrefix(r.UserAgent(), prefix) {
			t.Errorf("User-Agent must start with '%s', got '%s'", prefix, r.UserAgent())
		}
		w.Write([]byte("{}"))
	}))
	defer server.Close()
	conn, err := NewConnection(ConnectionSettings{
		URL:              server.URL,
		Beatname:         "testbeat",
		UserAgentPostfix: "Agent",
	})
	require.NoError(t, err)
	require.NoError(t, conn.Connect(), "conn.Connect must not return an error")
```

This way we can put the test code in the server receiving the request to assert the User-Agent and we execute the whole code used to create the user agent and make the request without any mock.

Reply:
I'm not sure that'll work without making the client throw an error? The client expects a response when it hits the `/` endpoint, and the reason for using the mock framework was because it would return "expected" responses, so the client will behave normally.

Reply:
It does work, I tested it before writing my suggestion and tested it again today.
```
tiago@millennium-falcon beats/libbeat/esleg/eslegclient  v1.22.2  beats-user-agent-managed [$!?] % go test -v -run=TestUserAgentHeader
=== RUN   TestUserAgentHeader
--- PASS: TestUserAgentHeader (0.00s)
PASS
ok      github.com/elastic/beats/v7/libbeat/esleg/eslegclient   0.209s
tiago@millennium-falcon beats/libbeat/esleg/eslegclient  v1.22.2  beats-user-agent-managed [$!?] % 
```

The version ends up being an empty string, which logs some messages about it, but it does not break the client, nor returns an error.

I'd still prefer the version I suggested that does not rely on the behaviour of any library, and it is a much easier to understand.

Reply:
> is there something we can do to mock-es or the test to make it easier to parse?

By "easier to parse" do you mean easier to understand by us, humans?

I'm biased to answer that as I believe we can make the test look like what I suggested on https://github.com/elastic/beats/pull/39403#discussion_r1638208533.

I believe that what currently makes the test hard to understand is that we want to test if the User-Agent is correctly set on a request, but to assert it rely on a side-effect from `mock-es`. We rely that the User-Agent will be present in the metrics, we don't even check if the string `testbeat-Agent` is an User-Agent, we pass the test if is there, no matter where.

Also if the test fails we log all the metrics, which is way more information than what we need to understand the failure.

By using the `httptest` package I suggested we look directly at the User-Agent our client is sending to the server, if the test fails we print only the User-Agent sent by our client. It seems much more straight forward to me.

Reply:
> We rely that the User-Agent will be present in the metrics, we don't even check if the string testbeat-Agent is an User-Agent, we pass the test if is there, no matter where.

I definitely agree there, the metrics don't really do much to separate out the user-agent data from the rest of the metrics, which is a tad confusing. Unfortunately the imported metrics library in `mock-es` doesn't really namespace things that way, so `user-agent.user_agent.Elastic-testbeat-Agent/8.15.0` is a string, not a nested map. I've reverted to your solution, since we don't have to do weird string manipulation that way.